bootloader/lkenv/lkenv.go

```diff
@@ -0,0 +1,278 @@
+				 - snap_mode, snap_try_kernel, snap_try_core behaves same way as with u-boot
+				 - boot partition labels are never modified by snapd at run time
+		 bootloader:
+				 - Finds boot partition to use based on info in matrix and snap_kernel / snap_try_kernel
+				 - bootloaer does not alter matrix, only alters snap_mode
+
+		 [ <bootimg 1 part label> ] [ <currently installed kernel snap revison> ]
+		 [ <bootimg 2 part label> ] [ <currently installed kernel snap revision> ]
+	*/
+	Bootimg_matrix [SNAP_BOOTIMG_PART_NUM][2][SNAP_NAME_MAX_LEN]byte
+	/* crc32 value for structure */
+	Crc32 uint32
+}
+
+// Env contains the data of the uboot environment
+// path can be file or partition device node
+type Env struct {
+	path string
+	pathbak string
+	env SnapBootSelect
+}
+
+//helper function to trim string from byte array to actual length
+func cToGoString(c []byte) string {
+	n := -1
+	for i, b := range c {
+		if b == 0 {
+			break
+		}
+		n = i
+	}
+	return string(c[:n+1])
+}
+
+// helper function to copy string into array and making sure it's terminated
+func copyString(b []byte, s string) {
+	if (s == "") {
+		b[0] = 0
+	} else {
+		copy(b[:], s)
+		b[len(s)] = 0
+	}
+}
+
+func NewEnv(path string) *Env {
+	// osutil.FileExists(path + "bak")
+	return &Env{
+		path: path,
+		pathbak: path + "bak",
+	}
+}
+
+func (l *Env) Get(key string) string {
+	switch key {
+	case "snap_mode":
+		return cToGoString(l.env.Snap_mode[:])
+	case "snap_kernel":
+		return cToGoString(l.env.Snap_kernel[:])
+	case "snap_try_kernel":
+		return cToGoString(l.env.Snap_try_kernel[:])
+	case "snap_core":
+		return cToGoString(l.env.Snap_core[:])
+	case "snap_try_core":
+		return cToGoString(l.env.Snap_try_core[:])
+	}
+	return ""
+}
+
+func (l *Env) Set(key, value string) {
+	switch key {
+	case "snap_mode":
+		copyString(l.env.Snap_mode[:], value)
+	case "snap_kernel":
+		copyString(l.env.Snap_kernel[:], value)
+	case "snap_try_kernel":
+		copyString(l.env.Snap_try_kernel[:], value)
+	case "snap_core":
+		copyString(l.env.Snap_core[:], value)
+	case "snap_try_core":
+		copyString(l.env.Snap_try_core[:], value)
+	}
+}
+
+// Configure partition labels for used boot partitions
+// this should not be used at run time, it should be used
+// once at image build time, if part labels are not filled by gadget build
+func (l *Env) ConfigureBootPartitions(boot_1, boot_2 string) {
+	copyString(l.env.Bootimg_matrix[0][0][:], boot_1)
+	copyString(l.env.Bootimg_matrix[1][0][:], boot_2)
+}
+
+func (l *Env) Load() error {
+	err := l.LoadEnv(l.path)
+	if (err != nil) {
+		return l.LoadEnv(l.pathbak)
+	}
+	return nil
+}
+
+func (l *Env) LoadEnv(path string) error {
+	f, err := os.Open(path)
+	if err != nil {
+		return fmt.Errorf("error loading lk environment file: %v", err)
+	}
+
+	defer f.Close()
+	if err := binary.Read(f, binary.LittleEndian, &l.env); err != nil {
+		return fmt.Errorf("error reading lk environment from file %v", err)
```

Comment:
fmt.Errorf("cannot read environment file: %v", err)

Reply:
Fixed

Reply:
not pushed yet?